### Imports and Settings

In [66]:
from sqlalchemy import create_engine
import pymysql
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Obtain Raw Data  
The Credit One customer data is on the MySQL sever. So in this stage I have to connect to the server, query the database for the customer data, and set it up as a Pandas dataframe, and then export the raw data as a .csv file. 

## Connect to MySQL & Query Database

In [67]:
#connect python to MySQL server (database?)
db_connection_str = 'mysql+pymysql://deepanalytics:Sqltask1234!@34.73.222.197/deepanalytics'

In [68]:
#create engine for SQLalchemy to interface with the database API (see notes below)
db_connection = create_engine(db_connection_str)

# first time successfully queried db without this line, so not entirely
#sure why it is needed. asked on slack, answer = link to oriely article


In [69]:
#query the credit one data and extract it into a pd df
sql_query = pd.read_sql('SELECT * FROM credit', con=db_connection)

In [70]:
#create pandas df from SQL query
df = pd.DataFrame(sql_query)

*Notes*
* First time through the exercise, I missed the ```db_connection = create_engine(db_connection_str)``` line. i got a name error b/c 'db_connection' was not defined. so i removed the _str from the first line, ran the cells, and was able to execute the query in my notebook. So ultimately, I'm unsure why we need the create_engine code. (see my SQL collection in edge for some explanatory links)
* in the course outline, they did not have us use ```pd.DataFrame()```. They just told us to create the df from ```df = pd.read_sql()```. I suspect that was an oversight.
* I made the slight changes you see following a tutorial I found on how to export a SQL query as a csv file.

## Verify Proper Extraction

In [71]:
df.head()

MyUnknownColumn         X1      X2          X3        X4   X5     X6     X7  \
0              ID  LIMIT_BAL     SEX   EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2   
1               1      20000  female  university         1   24      2      2   
2               2     120000  female  university         2   26     -1      2   
3               3      90000  female  university         2   34      0      0   
4               4      50000  female  university         1   37      0      0   

      X8     X9  ...        X15        X16        X17       X18       X19  \
0  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2   
1     -1     -1  ...          0          0          0         0       689   
2      0      0  ...       3272       3455       3261         0      1000   
3      0      0  ...      14331      14948      15549      1518      1500   
4      0      0  ...      28314      28959      29547      2000      2019   

        X20       X21       X22       X23                           Y  
0  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
1         0         0         0         0                     default  
2      1000      1000         0      2000                     default  
3      1000      1000      1000      5000                 not default  
4      1200      1100      1069      1000                 not default  

[5 rows x 25 columns]

In [72]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30204 entries, 0 to 30203
Data columns (total 25 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   MyUnknownColumn  30204 non-null  object
 1   X1               30204 non-null  object
 2   X2               30204 non-null  object
 3   X3               30204 non-null  object
 4   X4               30204 non-null  object
 5   X5               30204 non-null  object
 6   X6               30204 non-null  object
 7   X7               30204 non-null  object
 8   X8               30204 non-null  object
 9   X9               30204 non-null  object
 10  X10              30204 non-null  object
 11  X11              30204 non-null  object
 12  X12              30204 non-null  object
 13  X13              30204 non-null  object
 14  X14              30204 non-null  object
 15  X15              30204 non-null  object
 16  X16              30204 non-null  object
 17  X17              30204 non-null

## Export Raw Data As CSV File

In [73]:
#export df as .csv file
df.to_csv (r'C:\Users\Kpiat\export_data.csv', index = False)

*Notes*  

* This data is from a paper published in Expert Systems with Applications found [here](https://bradzzz.gitbooks.io/ga-seattle-dsi/content/dsi/dsi_05_classification_databases/2.1-lesson/assets/datasets/DefaultCreditCardClients_yeh_2009.pdf)
* A short summary of project and description of variables is in PDF in course 2 folder.
* Dataframe Characteristics: 26 columns—23 features/independent variables, 1 dependent variable, column 0 = ID, and 30,204 rows 
* Initial Cleaning Tasks:
1. Reset column headings/names to strings in row 0 
2. Delete row 0 from the dataframe.
3. Change dtypes for variables for object to appropriate type
4. First 200+ rows may be duplicates, investigate furthur. 

---

# Process Raw Data

In [74]:
#import exported csv file with raw data
df = pd.read_csv('raw_credit_one_data.csv')

In [75]:
#verify imported data looks as expected 
df.head()

MyUnknownColumn         X1      X2          X3        X4   X5     X6     X7  \
0              ID  LIMIT_BAL     SEX   EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2   
1               1      20000  female  university         1   24      2      2   
2               2     120000  female  university         2   26     -1      2   
3               3      90000  female  university         2   34      0      0   
4               4      50000  female  university         1   37      0      0   

      X8     X9  ...        X15        X16        X17       X18       X19  \
0  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2   
1     -1     -1  ...          0          0          0         0       689   
2      0      0  ...       3272       3455       3261         0      1000   
3      0      0  ...      14331      14948      15549      1518      1500   
4      0      0  ...      28314      28959      29547      2000      2019   

        X20       X21       X22       X23                           Y  
0  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
1         0         0         0         0                     default  
2      1000      1000         0      2000                     default  
3      1000      1000      1000      5000                 not default  
4      1200      1100      1069      1000                 not default  

[5 rows x 25 columns]

In [76]:
#look at number of rows and columns in df as benchmark before making any changes 
df.shape

(30204, 25)

## Change Column Headings 

In [77]:
#change column headings to values in index (row) 0 
df.columns = df.iloc[0]

In [78]:
#verify change was successful
df.head()

0  ID  LIMIT_BAL     SEX   EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  \
0  ID  LIMIT_BAL     SEX   EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3   
1   1      20000  female  university         1   24      2      2     -1   
2   2     120000  female  university         2   26     -1      2      0   
3   3      90000  female  university         2   34      0      0      0   
4   4      50000  female  university         1   37      0      0      0   

0  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3   
1     -1  ...          0          0          0         0       689         0   
2      0  ...       3272       3455       3261         0      1000      1000   
3      0  ...      14331      14948      15549      1518      1500      1000   
4      0  ...      28314      28959      29547      2000      2019      1200   

0  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
1         0         0         0                     default  
2      1000         0      2000                     default  
3      1000      1000      5000                 not default  
4      1100      1069      1000                 not default  

[5 rows x 25 columns]

In [79]:
#drop row 0 that contains column headings
df.drop(df.index[0], inplace = True)

In [80]:
#verify df is -1 row
len(df)

30203

In [81]:
#verify column headings row was dropped
df.head()

0 ID LIMIT_BAL     SEX   EDUCATION MARRIAGE AGE PAY_0 PAY_2 PAY_3 PAY_4  ...  \
1  1     20000  female  university        1  24     2     2    -1    -1  ...   
2  2    120000  female  university        2  26    -1     2     0     0  ...   
3  3     90000  female  university        2  34     0     0     0     0  ...   
4  4     50000  female  university        1  37     0     0     0     0  ...   
5  5     50000    male  university        1  57    -1     0    -1     0  ...   

0 BILL_AMT4 BILL_AMT5 BILL_AMT6 PAY_AMT1 PAY_AMT2 PAY_AMT3 PAY_AMT4 PAY_AMT5  \
1         0         0         0        0      689        0        0        0   
2      3272      3455      3261        0     1000     1000     1000        0   
3     14331     14948     15549     1518     1500     1000     1000     1000   
4     28314     28959     29547     2000     2019     1200     1100     1069   
5     20940     19146     19131     2000    36681    10000     9000      689   

0 PAY_AMT6 default payment next month  
1        0                    default  
2     2000                    default  
3     5000                not default  
4     1000                not default  
5      679                not default  

[5 rows x 25 columns]

*Notes*
* Found how to do this on [Stack Overflow](https://stackoverflow.com/questions/26147180/convert-row-to-column-header-for-pandas-dataframe).
* ```df.iloc[pd.RangeIndex(len(df)).drop(0)]``` is for situations where your df does not have or you don't know if it has unique index numbers for each row. If you know all index numbers are unique, you can use ```df.drop(df.index[0])```.
* I choose non-unique method because I'm unsure what the situation is. 

***♠ Why is the 0 index missing? I thought df.drop would drop data from index 0 and replace it with values that used to be in index 1.***

In [82]:
df.loc[200:206]

0     ID  LIMIT_BAL     SEX        EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  \
200  200      30000  female      high school         2   22      1      2   
201  201     180000  female  graduate school         1   38     -2     -2   
202  NaN         X1      X2               X3        X4   X5     X6     X7   
203   ID  LIMIT_BAL     SEX        EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2   
204    1      20000  female       university         1   24      2      2   
205    2     120000  female       university         2   26     -1      2   
206    3      90000  female       university         2   34      0      0   

0    PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
200      2      0  ...      29836       1630          0      1000        85   
201     -2     -2  ...          0          0          0         0         0   
202     X8     X9  ...        X15        X16        X17       X18       X19   
203  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2   
204     -1     -1  ...          0          0          0         0       689   
205      0      0  ...       3272       3455       3261         0      1000   
206      0      0  ...      14331      14948      15549      1518      1500   

0    PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
200      1714       104         0         0                     default  
201         0         0         0         0                 not default  
202       X20       X21       X22       X23                           Y  
203  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
204         0         0         0         0                     default  
205      1000      1000         0      2000                     default  
206      1000      1000      1000      5000                 not default  

[7 rows x 25 columns]

## Wrangle Duplicate Data

***♠ How do we look for duplicate columns? Do we need to do that? How would we drop duplicate columns?***

### Duplicate Rows

In [83]:
#find and print all duplicate rows in the df
duplicate_rows = df[df.duplicated()]
print(duplicate_rows)

0     ID LIMIT_BAL     SEX        EDUCATION MARRIAGE AGE PAY_0 PAY_2 PAY_3  \
204    1     20000  female       university        1  24     2     2    -1   
205    2    120000  female       university        2  26    -1     2     0   
206    3     90000  female       university        2  34     0     0     0   
207    4     50000  female       university        1  37     0     0     0   
208    5     50000    male       university        1  57    -1     0    -1   
..   ...       ...     ...              ...      ...  ..   ...   ...   ...   
400  197    150000  female       university        1  34    -2    -2    -2   
401  198     20000  female  graduate school        2  22     0     0     0   
402  199    500000  female  graduate school        1  34    -2    -2    -2   
403  200     30000  female      high school        2  22     1     2     2   
404  201    180000  female  graduate school        1  38    -2    -2    -2   

0   PAY_4  ... BILL_AMT4 BILL_AMT5 BILL_AMT6 PAY_AMT1 PAY_AMT2 

In [84]:
#drop duplicate rows from the df
df.drop_duplicates(inplace = True)  #inplace=True is to make change to df permanent

In [85]:
#verify rows were dropped
len(df)

30002

## Missing Data

In [86]:
#check for missing values
print(df.isnull().sum())

0
ID                            1
LIMIT_BAL                     0
SEX                           0
EDUCATION                     0
MARRIAGE                      0
AGE                           0
PAY_0                         0
PAY_2                         0
PAY_3                         0
PAY_4                         0
PAY_5                         0
PAY_6                         0
BILL_AMT1                     0
BILL_AMT2                     0
BILL_AMT3                     0
BILL_AMT4                     0
BILL_AMT5                     0
BILL_AMT6                     0
PAY_AMT1                      0
PAY_AMT2                      0
PAY_AMT3                      0
PAY_AMT4                      0
PAY_AMT5                      0
PAY_AMT6                      0
default payment next month    0
dtype: int64


*Notes*
* There appears to be one one missing data point in the ID var.
* However, I'm not sure running ```df.isnull()``` is enough to check for that.

In [87]:
#drop the row with missing ID
df.dropna(inplace=True)

In [88]:
len(df)

30001

---

## Invaild Data

In [89]:
df.head()

0 ID LIMIT_BAL     SEX   EDUCATION MARRIAGE AGE PAY_0 PAY_2 PAY_3 PAY_4  ...  \
1  1     20000  female  university        1  24     2     2    -1    -1  ...   
2  2    120000  female  university        2  26    -1     2     0     0  ...   
3  3     90000  female  university        2  34     0     0     0     0  ...   
4  4     50000  female  university        1  37     0     0     0     0  ...   
5  5     50000    male  university        1  57    -1     0    -1     0  ...   

0 BILL_AMT4 BILL_AMT5 BILL_AMT6 PAY_AMT1 PAY_AMT2 PAY_AMT3 PAY_AMT4 PAY_AMT5  \
1         0         0         0        0      689        0        0        0   
2      3272      3455      3261        0     1000     1000     1000        0   
3     14331     14948     15549     1518     1500     1000     1000     1000   
4     28314     28959     29547     2000     2019     1200     1100     1069   
5     20940     19146     19131     2000    36681    10000     9000      689   

0 PAY_AMT6 default payment next month  
1        0                    default  
2     2000                    default  
3     5000                not default  
4     1000                not default  
5      679                not default  

[5 rows x 25 columns]

In [90]:
#locate rows 201-206 (not inclusive)
df.loc[201:206]

0     ID  LIMIT_BAL     SEX        EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  \
201  201     180000  female  graduate school         1   38     -2     -2   
203   ID  LIMIT_BAL     SEX        EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2   

0    PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
201     -2     -2  ...          0          0          0         0         0   
203  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2   

0    PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
201         0         0         0         0                 not default  
203  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  

[2 rows x 25 columns]

In [91]:
#drop row 203 containing original column headings
df.drop(203, inplace=True)

In [92]:
len(df)

30000

In [93]:
df.loc[201:210] #only row 201 should appear b/c others were dropped above

0     ID LIMIT_BAL     SEX        EDUCATION MARRIAGE AGE PAY_0 PAY_2 PAY_3  \
201  201    180000  female  graduate school        1  38    -2    -2    -2   

0   PAY_4  ... BILL_AMT4 BILL_AMT5 BILL_AMT6 PAY_AMT1 PAY_AMT2 PAY_AMT3  \
201    -2  ...         0         0         0        0        0        0   

0   PAY_AMT4 PAY_AMT5 PAY_AMT6 default payment next month  
201        0        0        0                not default  

[1 rows x 25 columns]

*Notes*
* deleting column headings row and duplicate rows took many tries.
    * First problem, I didn't know that dropping rows doesn't stick unless you add an inplace = True parameter to your code. 
    * Second problem, I didn't realize that after you drop an index from the df (and stick it) that the index number dropped no longer appears in the df (e.g. if you drop index 0, when you use df.head() the first entry will be index 1). In other words, the data from below does not move up and fill the dropped indices as it would in an Excel file. 
* After I noticed the drops weren't sticking, I tried to solve the problem by naming a new df and setting that = to the df.drop_duplicated. That was unsuccessful. 
* Lavana advised me to add an 'inplace=True' parameter to all my drop commands. [this](https://stackabuse.com/python-with-pandas-dataframe-tutorial-with-examples/) helped me understand why that parameter is needed.
* I found this code on a tutorial to find non-numeric data in a column ```non_numeric_limit = df.LIMIT_BAL.str.contains('[^0-9.-]') / df.loc[non_numeric_limit].head()```
---

## Re-set Data Types of Variables

In [94]:
df.head()

0 ID LIMIT_BAL     SEX   EDUCATION MARRIAGE AGE PAY_0 PAY_2 PAY_3 PAY_4  ...  \
1  1     20000  female  university        1  24     2     2    -1    -1  ...   
2  2    120000  female  university        2  26    -1     2     0     0  ...   
3  3     90000  female  university        2  34     0     0     0     0  ...   
4  4     50000  female  university        1  37     0     0     0     0  ...   
5  5     50000    male  university        1  57    -1     0    -1     0  ...   

0 BILL_AMT4 BILL_AMT5 BILL_AMT6 PAY_AMT1 PAY_AMT2 PAY_AMT3 PAY_AMT4 PAY_AMT5  \
1         0         0         0        0      689        0        0        0   
2      3272      3455      3261        0     1000     1000     1000        0   
3     14331     14948     15549     1518     1500     1000     1000     1000   
4     28314     28959     29547     2000     2019     1200     1100     1069   
5     20940     19146     19131     2000    36681    10000     9000      689   

0 PAY_AMT6 default payment next month  
1        0                    default  
2     2000                    default  
3     5000                not default  
4     1000                not default  
5      679                not default  

[5 rows x 25 columns]

In [95]:
#look at data types for each variable
df.dtypes

0
ID                            object
LIMIT_BAL                     object
SEX                           object
EDUCATION                     object
MARRIAGE                      object
AGE                           object
PAY_0                         object
PAY_2                         object
PAY_3                         object
PAY_4                         object
PAY_5                         object
PAY_6                         object
BILL_AMT1                     object
BILL_AMT2                     object
BILL_AMT3                     object
BILL_AMT4                     object
BILL_AMT5                     object
BILL_AMT6                     object
PAY_AMT1                      object
PAY_AMT2                      object
PAY_AMT3                      object
PAY_AMT4                      object
PAY_AMT5                      object
PAY_AMT6                      object
default payment next month    object
dtype: object

*Notes*
* I thought after dropping all non-numerical values from the df, the vars with numerical values would change to int64 or float64. They did not. Ben said data types will need to be changed manually.

### Continuous Variables

In [96]:
df.LIMIT_BAL.value_counts()

50000      3365
20000      1976
30000      1610
80000      1567
200000     1528
           ... 
800000        2
690000        1
760000        1
327680        1
1000000       1
Name: LIMIT_BAL, Length: 81, dtype: int64

In [97]:
non_numeric_limit = df.LIMIT_BAL.str.contains('[^0-9.-]') 
df.loc[non_numeric_limit].head()

Empty DataFrame
Columns: [ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6, default payment next month]
Index: []

[0 rows x 25 columns]

In [98]:
#change data type of LIMIT_BAL var from object to int
df["LIMIT_BAL"] = df['LIMIT_BAL'].astype('int')

In [99]:
df.AGE.value_counts()

29    1605
27    1477
28    1409
30    1395
26    1256
31    1217
25    1186
34    1162
32    1158
33    1146
24    1127
35    1113
36    1108
37    1041
39     954
38     944
23     931
40     870
41     824
42     794
44     700
43     670
45     617
46     570
22     560
47     501
48     466
49     452
50     411
51     340
53     325
52     304
54     247
55     209
56     178
58     122
57     122
59      83
21      67
60      67
61      56
62      44
63      31
64      31
66      25
65      24
67      16
69      15
70      10
68       5
73       4
75       3
71       3
72       3
74       1
79       1
Name: AGE, dtype: int64

In [100]:
#change data type of AGE var from object to int
df["AGE"] = df['AGE'].astype('int')

In [101]:
df.PAY_0.value_counts()

0     14737
-1     5686
1      3688
-2     2759
2      2667
3       322
4        76
5        26
8        19
6        11
7         9
Name: PAY_0, dtype: int64

In [102]:
df.PAY_4.value_counts()

0     16455
-1     5687
-2     4348
2      3159
3       180
4        69
7        58
5        35
6         5
1         2
8         2
Name: PAY_4, dtype: int64

In [103]:
df.BILL_AMT1.value_counts()

0         2008
390        244
780         76
326         72
316         63
          ... 
209396       1
19189        1
2422         1
182441       1
57889        1
Name: BILL_AMT1, Length: 22723, dtype: int64

In [104]:
#change data type of BILL_AMT1 var from object to int
df["BILL_AMT1"] = df['BILL_AMT1'].astype('int')

In [105]:
#change data type of BILL_AMT2 var from object to int
df["BILL_AMT2"] = df['BILL_AMT2'].astype('int')

In [106]:
#change data type of BILL_AMT3 var from object to int
df["BILL_AMT3"] = df['BILL_AMT3'].astype('int')

In [107]:
#change data type of BILL_AMT4 var from object to int
df["BILL_AMT4"] = df['BILL_AMT4'].astype('int')

In [108]:
#change data type of BILL_AMT5 var from object to int
df["BILL_AMT5"] = df['BILL_AMT5'].astype('int')

In [109]:
#change data type of BILL_AMT6 var from object to int
df["BILL_AMT6"] = df['BILL_AMT6'].astype('int')

In [110]:
df.PAY_AMT4.value_counts()

0        6408
1000     1394
2000     1214
3000      887
5000      810
         ... 
1599        1
86417       1
5577        1
3063        1
11568       1
Name: PAY_AMT4, Length: 6937, dtype: int64

In [111]:
#change data type of PAY_AMT1 var from object to int
df["PAY_AMT1"] = df['PAY_AMT1'].astype('int')

In [112]:
#change data type of PAY_AMT2 var from object to int
df["PAY_AMT2"] = df['PAY_AMT2'].astype('int')

In [113]:
#change data type of PAY_AMT3 var from object to int
df["PAY_AMT3"] = df['PAY_AMT3'].astype('int')

In [114]:
#change data type of PAY_AMT4 var from object to int
df["PAY_AMT4"] = df['PAY_AMT4'].astype('int')

In [115]:
#change data type of PAY_AMT5 var from object to int
df["PAY_AMT5"] = df['PAY_AMT5'].astype('int')

In [116]:
#change data type of PAY_AMT6 var from object to int
df["PAY_AMT6"] = df['PAY_AMT6'].astype('int')

In [117]:
#rename default payment next month column
df.rename(columns={"default payment next month": "DEFAULT"}, inplace = True)

In [118]:
#verify changes were made
df.dtypes

0
ID           object
LIMIT_BAL     int32
SEX          object
EDUCATION    object
MARRIAGE     object
AGE           int32
PAY_0        object
PAY_2        object
PAY_3        object
PAY_4        object
PAY_5        object
PAY_6        object
BILL_AMT1     int32
BILL_AMT2     int32
BILL_AMT3     int32
BILL_AMT4     int32
BILL_AMT5     int32
BILL_AMT6     int32
PAY_AMT1      int32
PAY_AMT2      int32
PAY_AMT3      int32
PAY_AMT4      int32
PAY_AMT5      int32
PAY_AMT6      int32
DEFAULT      object
dtype: object

### Categorical Variables

In [119]:
df.SEX.value_counts()

female    18112
male      11888
Name: SEX, dtype: int64

In [120]:
df.EDUCATION.value_counts()

university         14030
graduate school    10585
high school         4917
other                468
Name: EDUCATION, dtype: int64

In [121]:
df.MARRIAGE.value_counts()

2    15964
1    13659
3      323
0       54
Name: MARRIAGE, dtype: int64

In [122]:
df.PAY_0.value_counts()

0     14737
-1     5686
1      3688
-2     2759
2      2667
3       322
4        76
5        26
8        19
6        11
7         9
Name: PAY_0, dtype: int64

In [123]:
df.DEFAULT.value_counts()

not default    23364
default         6636
Name: DEFAULT, dtype: int64

# Export Cleaned Data

In [124]:
df.to_csv (r'C:\Users\kpiat\Data Science Stuff\Course 2-Data Program\cleaned_credit_one_data.csv', index = False, header=True)
print(df)

0         ID  LIMIT_BAL     SEX    EDUCATION MARRIAGE  AGE PAY_0 PAY_2 PAY_3  \
1          1      20000  female   university        1   24     2     2    -1   
2          2     120000  female   university        2   26    -1     2     0   
3          3      90000  female   university        2   34     0     0     0   
4          4      50000  female   university        1   37     0     0     0   
5          5      50000    male   university        1   57    -1     0    -1   
...      ...        ...     ...          ...      ...  ...   ...   ...   ...   
30199  29996     220000    male  high school        1   39     0     0     0   
30200  29997     150000    male  high school        2   43    -1    -1    -1   
30201  29998      30000    male   university        2   37     4     3     2   
30202  29999      80000    male  high school        1   41     1    -1     0   
30203  30000      50000    male   university        1   46     0     0     0   

0     PAY_4  ... BILL_AMT4 BILL_AMT5  B

# Transform Data

## Rename Features

In [125]:
#rename PAY features to convey month
df.rename(columns = {'PAY_0': 'pay_s', 'PAY_2':'pay_ag', 'PAY_3':'pay_jy','PAY_4':'pay_ju', 'PAY_5':'pay_m', 'PAY_6':'pay_ap'}, inplace = True)

In [127]:
#rename BILL_AMT features to convey month
df.rename(columns = {'BILL_AMT1': 'bill_s', 'BILL_AMT2':'bill_ag', 'BILL_AMT3':'bill_jy','BILL_AMT4':'bill_ju', 'BILL_AMT5':'bill_m', 'BILL_AMT6':'bill_ap'}, inplace = True)

In [128]:
#rename PAY_AMT features to convey month
df.rename(columns = {'PAY_AMT1': 'pmt_s', 'PAY_AMT2':'pmt_ag', 'PAY_AMT3':'pmt_jy','PAY_AMT4':'pmt_ju', 'PAY_AMT5':'pmt_m', 'PAY_AMT6':'pmt_ap'}, inplace = True)

In [129]:
#verify features were renamed
df.head()

0 ID  LIMIT_BAL     SEX   EDUCATION MARRIAGE  AGE pay_s pay_ag pay_jy pay_ju  \
1  1      20000  female  university        1   24     2      2     -1     -1   
2  2     120000  female  university        2   26    -1      2      0      0   
3  3      90000  female  university        2   34     0      0      0      0   
4  4      50000  female  university        1   37     0      0      0      0   
5  5      50000    male  university        1   57    -1      0     -1      0   

0  ... bill_ju bill_m  bill_ap  pmt_s  pmt_ag  pmt_jy  pmt_ju  pmt_m  pmt_ap  \
1  ...       0      0        0      0     689       0       0      0       0   
2  ...    3272   3455     3261      0    1000    1000    1000      0    2000   
3  ...   14331  14948    15549   1518    1500    1000    1000   1000    5000   
4  ...   28314  28959    29547   2000    2019    1200    1100   1069    1000   
5  ...   20940  19146    19131   2000   36681   10000    9000    689     679   

0      DEFAULT  
1      default  
2      default  
3  not default  
4  not default  
5  not default  

[5 rows x 25 columns]

## Reorder Features

In [130]:
#define a new fuction to move and re-order columns in a df (copied from toward data science post)
def movecol(df, cols_to_move=[], ref_col='', place='After'):
    
    cols = df.columns.tolist()
    if place == 'After':
        seg1 = cols[:list(cols).index(ref_col) + 1]
        seg2 = cols_to_move
    if place == 'Before':
        seg1 = cols[:list(cols).index(ref_col)]
        seg2 = cols_to_move + [ref_col]
    
    seg1 = [i for i in seg1 if i not in seg2]
    seg3 = [i for i in cols if i not in seg1 + seg2]
    
    return(df[seg1 + seg2 + seg3])

In [132]:
#move LIMIT_BAL column next to other account/credit variables
df = movecol(df, 
             cols_to_move=['LIMIT_BAL'], 
             ref_col='AGE',
             place='After')
df

0         ID     SEX    EDUCATION MARRIAGE  AGE  LIMIT_BAL pay_s pay_ag  \
1          1  female   university        1   24      20000     2      2   
2          2  female   university        2   26     120000    -1      2   
3          3  female   university        2   34      90000     0      0   
4          4  female   university        1   37      50000     0      0   
5          5    male   university        1   57      50000    -1      0   
...      ...     ...          ...      ...  ...        ...   ...    ...   
30199  29996    male  high school        1   39     220000     0      0   
30200  29997    male  high school        2   43     150000    -1     -1   
30201  29998    male   university        2   37      30000     4      3   
30202  29999    male  high school        1   41      80000     1     -1   
30203  30000    male   university        1   46      50000     0      0   

0     pay_jy pay_ju  ... bill_ju bill_m  bill_ap  pmt_s  pmt_ag  pmt_jy  \
1         -1     -1  ...       0      0        0      0     689       0   
2          0      0  ...    3272   3455     3261      0    1000    1000   
3          0      0  ...   14331  14948    15549   1518    1500    1000   
4          0      0  ...   28314  28959    29547   2000    2019    1200   
5         -1      0  ...   20940  19146    19131   2000   36681   10000   
...      ...    ...  ...     ...    ...      ...    ...     ...     ...   
30199      0      0  ...   88004  31237    15980   8500   20000    5003   
30200     -1     -1  ...    8979   5190        0   1837    3526    8998   
30201      2     -1  ...   20878  20582    19357      0       0   22000   
30202      0      0  ...   52774  11855    48944  85900    3409    1178   
30203      0      0  ...   36535  32428    15313   2078    1800    1430   

0      pmt_ju  pmt_m  pmt_ap      DEFAULT  
1           0      0       0      default  
2        1000      0    2000      default  
3        1000   1000    5000  not default  
4        1100   1069    1000  not default  
5        9000    689     679  not default  
...       ...    ...     ...          ...  
30199    3047   5000    1000  not default  
30200     129      0       0  not default  
30201    4200   2000    3100      default  
30202    1926  52964    1804      default  
30203    1000   1000    1000      default  

[30000 rows x 25 columns]

In [134]:
#re-order pay columns
df = movecol(df, 
             cols_to_move=['pay_ap', 'pay_m', 'pay_ju', 'pay_jy', 'pay_ag', 'pay_s'], 
             ref_col='LIMIT_BAL',
             place='After')
df

0         ID     SEX    EDUCATION MARRIAGE  AGE  LIMIT_BAL pay_ap pay_m  \
1          1  female   university        1   24      20000     -2    -2   
2          2  female   university        2   26     120000      2     0   
3          3  female   university        2   34      90000      0     0   
4          4  female   university        1   37      50000      0     0   
5          5    male   university        1   57      50000      0     0   
...      ...     ...          ...      ...  ...        ...    ...   ...   
30199  29996    male  high school        1   39     220000      0     0   
30200  29997    male  high school        2   43     150000      0     0   
30201  29998    male   university        2   37      30000      0     0   
30202  29999    male  high school        1   41      80000     -1     0   
30203  30000    male   university        1   46      50000      0     0   

0     pay_ju pay_jy  ... bill_ju bill_m  bill_ap  pmt_s  pmt_ag  pmt_jy  \
1         -1     -1  ...       0      0        0      0     689       0   
2          0      0  ...    3272   3455     3261      0    1000    1000   
3          0      0  ...   14331  14948    15549   1518    1500    1000   
4          0      0  ...   28314  28959    29547   2000    2019    1200   
5          0     -1  ...   20940  19146    19131   2000   36681   10000   
...      ...    ...  ...     ...    ...      ...    ...     ...     ...   
30199      0      0  ...   88004  31237    15980   8500   20000    5003   
30200     -1     -1  ...    8979   5190        0   1837    3526    8998   
30201     -1      2  ...   20878  20582    19357      0       0   22000   
30202      0      0  ...   52774  11855    48944  85900    3409    1178   
30203      0      0  ...   36535  32428    15313   2078    1800    1430   

0      pmt_ju  pmt_m  pmt_ap      DEFAULT  
1           0      0       0      default  
2        1000      0    2000      default  
3        1000   1000    5000  not default  
4        1100   1069    1000  not default  
5        9000    689     679  not default  
...       ...    ...     ...          ...  
30199    3047   5000    1000  not default  
30200     129      0       0  not default  
30201    4200   2000    3100      default  
30202    1926  52964    1804      default  
30203    1000   1000    1000      default  

[30000 rows x 25 columns]

In [135]:
#re-order pmt columns
df = movecol(df, 
             cols_to_move=['pmt_ap', 'pmt_m', 'pmt_ju', 'pmt_jy', 'pmt_ag', 'pmt_s'], 
             ref_col='LIMIT_BAL',
             place='After')
df

0         ID     SEX    EDUCATION MARRIAGE  AGE  LIMIT_BAL  pmt_ap  pmt_m  \
1          1  female   university        1   24      20000       0      0   
2          2  female   university        2   26     120000    2000      0   
3          3  female   university        2   34      90000    5000   1000   
4          4  female   university        1   37      50000    1000   1069   
5          5    male   university        1   57      50000     679    689   
...      ...     ...          ...      ...  ...        ...     ...    ...   
30199  29996    male  high school        1   39     220000    1000   5000   
30200  29997    male  high school        2   43     150000       0      0   
30201  29998    male   university        2   37      30000    3100   2000   
30202  29999    male  high school        1   41      80000    1804  52964   
30203  30000    male   university        1   46      50000    1000   1000   

0      pmt_ju  pmt_jy  ...  pay_jy  pay_ag pay_s  bill_s bill_ag bill_jy  \
1           0       0  ...      -1       2     2    3913    3102     689   
2        1000    1000  ...       0       2    -1    2682    1725    2682   
3        1000    1000  ...       0       0     0   29239   14027   13559   
4        1100    1200  ...       0       0     0   46990   48233   49291   
5        9000   10000  ...      -1       0    -1    8617    5670   35835   
...       ...     ...  ...     ...     ...   ...     ...     ...     ...   
30199    3047    5003  ...       0       0     0  188948  192815  208365   
30200     129    8998  ...      -1      -1    -1    1683    1828    3502   
30201    4200   22000  ...       2       3     4    3565    3356    2758   
30202    1926    1178  ...       0      -1     1   -1645   78379   76304   
30203    1000    1430  ...       0       0     0   47929   48905   49764   

0     bill_ju bill_m  bill_ap      DEFAULT  
1           0      0        0      default  
2        3272   3455     3261      default  
3       14331  14948    15549  not default  
4       28314  28959    29547  not default  
5       20940  19146    19131  not default  
...       ...    ...      ...          ...  
30199   88004  31237    15980  not default  
30200    8979   5190        0  not default  
30201   20878  20582    19357      default  
30202   52774  11855    48944      default  
30203   36535  32428    15313      default  

[30000 rows x 25 columns]

In [136]:
#re-order bill columns
df = movecol(df, 
             cols_to_move=['bill_ap', 'bill_m', 'bill_ju', 'bill_jy', 'bill_ag', 'bill_s'], 
             ref_col='LIMIT_BAL',
             place='After')
df

0         ID     SEX    EDUCATION MARRIAGE  AGE  LIMIT_BAL  bill_ap  bill_m  \
1          1  female   university        1   24      20000        0       0   
2          2  female   university        2   26     120000     3261    3455   
3          3  female   university        2   34      90000    15549   14948   
4          4  female   university        1   37      50000    29547   28959   
5          5    male   university        1   57      50000    19131   19146   
...      ...     ...          ...      ...  ...        ...      ...     ...   
30199  29996    male  high school        1   39     220000    15980   31237   
30200  29997    male  high school        2   43     150000        0    5190   
30201  29998    male   university        2   37      30000    19357   20582   
30202  29999    male  high school        1   41      80000    48944   11855   
30203  30000    male   university        1   46      50000    15313   32428   

0      bill_ju  bill_jy  ...  pmt_jy  pmt_ag  pmt_s  pay_ap  pay_m  pay_ju  \
1            0      689  ...       0     689      0      -2     -2      -1   
2         3272     2682  ...    1000    1000      0       2      0       0   
3        14331    13559  ...    1000    1500   1518       0      0       0   
4        28314    49291  ...    1200    2019   2000       0      0       0   
5        20940    35835  ...   10000   36681   2000       0      0       0   
...        ...      ...  ...     ...     ...    ...     ...    ...     ...   
30199    88004   208365  ...    5003   20000   8500       0      0       0   
30200     8979     3502  ...    8998    3526   1837       0      0      -1   
30201    20878     2758  ...   22000       0      0       0      0      -1   
30202    52774    76304  ...    1178    3409  85900      -1      0       0   
30203    36535    49764  ...    1430    1800   2078       0      0       0   

0      pay_jy  pay_ag pay_s      DEFAULT  
1          -1       2     2      default  
2           0       2    -1      default  
3           0       0     0  not default  
4           0       0     0  not default  
5          -1       0    -1  not default  
...       ...     ...   ...          ...  
30199       0       0     0  not default  
30200      -1      -1    -1  not default  
30201       2       3     4      default  
30202       0      -1     1      default  
30203       0       0     0      default  

[30000 rows x 25 columns]

In [137]:
#export csv file with features renamed and reordered
df.to_csv (r'C:\Users\kpiat\Data Science Stuff\Course 2-Data Program\cleaned_transformed_credit_one_data.csv', index = False, header=True)

---